### PR TITLE
[Opmondev-153]: optimise correct process to use `xRequestId` 

### PR DESCRIPTION
**Pull request by wisecrow**
Corresponding changes:
1. Use `xRequestId` as matching criteria for `Client` `Producer` pairs.
2. If message has no `xRequestId` mark it as orphan and set `correctorStatus` to done.
3.  Keep `settings["corrector"]["documents-max"]` to fetch documents by batches, because `raw_messages` could have huge amount of documents and infrastructure could not handle such request.
4. Process timeout documents only having `xRequestId`
5. Remove `clientHash` and `producerHash` from clean data documents.
6. Set `xRequestId` in clean data documents.

Note: new indexes to MongoDB `clean_data` must be added:
https://www.mongodb.com/docs/manual/reference/method/db.collection.createIndex/

`{'xRequestId': 1}`,
`{'correctorStatus': 1, 'xRequestId': 1}`,
`{'client.xRequestId': 1}`,
`{'producer.xRequestId': 1}`,

- [ ]  Supplement documentation about changes
- [x] Add notes how to add indexes (`xRequestId`) to MongoDB

- [x] Fix `test_corrector_batch.py`

### CODE REVIEW

File: corrector_module/opmon_corrector/corrector_batch.py [hunk 1]
```diff
@@ -91,40 +114,76 @@ def _batch_run(self, process_dict):
...
+        while element_in_queue:
+            try:
+                element = to_remove_queue.get(False)
+                db_m.remove_duplicate_from_raw(element)
+                total_raw_removed += 1
+            except queue.Empty:
+                element_in_queue = False
+        if total_raw_removed > 0:
+            self.logger_m.log_info('corrector_batch_remove_duplicates_from_raw',
+                                   "Total of {0} duplicate documents removed from raw messages.".format(
```
Comment (justinask7) [inline corrector_module/opmon_corrector/corrector_batch.py:142]: Consistency for quotes

File: corrector_module/opmon_corrector/corrector_batch.py [hunk 2]
```diff
@@ -91,40 +114,76 @@ def _batch_run(self, process_dict):
...
-            # Configure worker
-            worker = CorrectorWorker(self.settings, f'worker_{i}')
-            p = multiprocessing.Process(target=worker.run, args=(list_to_process, duplicates))
-            pool.append(p)
+        # Go through the to_remove list and remove the duplicates
+        element_in_queue = True
+        total_raw_removed = 0
+        while element_in_queue:
+            try:
+                element = to_remove_queue.get(False)
```
Comment (justinask7) [inline corrector_module/opmon_corrector/corrector_batch.py:135]: `to_remove_queue.get(False)` seems quite odd. What is `to_remove_queue`? 🤔

File: corrector_module/opmon_corrector/corrector_batch.py [hunk 3]
```diff
@@ -91,40 +114,76 @@ def _batch_run(self, process_dict):
...
+
+        end_processing_time = time.time()
+        total_time = time.strftime("%H:%M:%S", time.gmtime(end_processing_time - start_processing_time))
+        msg = ["Number of duplicates: {0}".format(duplicates.value),
+               "Documents processed: " + str(doc_len),
+               "Processing time: {0}".format(total_time)]
+
+        self.logger_m.log_info('corrector_batch_end', ' | '.join(msg))
+        self.logger_m.log_heartbeat(
+            "finished", "SUCCEEDED")
```
Comment (justinask7) [inline corrector_module/opmon_corrector/corrector_batch.py:158]: line break should not be needed here

File: corrector_module/opmon_corrector/corrector_batch.py [hunk 4]
```diff
@@ -91,40 +114,76 @@ def _batch_run(self, process_dict):
...
-        # Starts all pool process
-        for p in pool:
-            p.start()
-        # Wait all processes to finish their jobs
-        for p in pool:
-            p.join()
+        doc_len += total_raw_removed
+
+        end_processing_time = time.time()
+        total_time = time.strftime("%H:%M:%S", time.gmtime(end_processing_time - start_processing_time))
```
Comment (justinask7) [inline corrector_module/opmon_corrector/corrector_batch.py:151]: Would it make sense to make the format a constant?

File: corrector_module/opmon_corrector/corrector_batch.py [hunk 5]
```diff
@@ -91,40 +114,76 @@ def _batch_run(self, process_dict):
...
+        process_dict['doc_len'] = doc_len
+
+        # Process documents without xRequestId
+        cursor = db_m.get_faulty_raw_documents()
+        self.logger_m.log_info('corrector_batch_raw', 'Processing {0} faulty raw documents'.format(len(cursor)))
+
+        for _doc in cursor:
+            fixed_doc = doc_m.correct_structure(_doc)
+            producer = fixed_doc if fixed_doc['securityServerType'] == 'Producer' else None
+            client = fixed_doc if fixed_doc['securityServerType'] == 'Client' else None
```
Comment (justinask7) [inline corrector_module/opmon_corrector/corrector_batch.py:168]: Let's move to the constants `Producer` and `Client`

**Review by justinask7 — APPROVED** [review]
main comments:
- Maybe would make sense to add flake8-quotes package to be sure about the quotes? in case it's hard to add it within the pipelines atm, we can just try to be sure to have it consistent within new PRs
- I'd strongly suggest start initiating constants, especially for repeatable hardcoded values

File: corrector_module/opmon_corrector/corrector_batch.py [hunk 6]
```diff
@@ -51,38 +50,62 @@ def run(self, process_dict):
...
+        Processes the workers in a thread pool.
+        :param list_to_process: queue of items to be processed by the worker processes
+        :param duplicates: a shared Value object to store the number of duplicates encountered during processing
+        : return: None
+        """
+        # Configure worker
+        pool = []
+        for i in range(self.settings["corrector"]["thread-count"]):
+            # Configure worker
+            worker = CorrectorWorker(self.settings, f'worker_{i}')
```
[the remark this answers is not in the document]
Comment (wisecrow) [reply corrector_module/opmon_corrector/corrector_batch.py]: Done, good point!

File: corrector_module/opmon_corrector/corrector_batch.py (hunk 2 above)
Comment (wisecrow) [reply corrector_module/opmon_corrector/corrector_batch.py:134]: `to_remove_queue` is `multiprocessing.Queue` object, which has `get()` and  `False` mean "return an item if one is immediately available, else raise the `queue.Empty` "
https://docs.python.org/3.9/library/multiprocessing.html#multiprocessing.Queue.get

File: corrector_module/opmon_corrector/corrector_batch.py (hunk 3 above)
Comment (wisecrow) [reply corrector_module/opmon_corrector/corrector_batch.py]: Fixed

File: corrector_module/opmon_corrector/corrector_batch.py (hunk 4 above)
Comment (wisecrow) [reply corrector_module/opmon_corrector/corrector_batch.py]: Yes, added a constant

File: corrector_module/opmon_corrector/corrector_batch.py (hunk 5 above)
Comment (wisecrow) [reply corrector_module/opmon_corrector/corrector_batch.py:171]: Done

File: corrector_module/opmon_corrector/corrector_batch.py [hunk 7]
```diff
@@ -91,40 +117,79 @@ def _batch_run(self, process_dict):
...
+        doc_len += total_raw_removed
+
+        end_processing_time = time.time()
+        total_time = time.strftime(
+            PROCESSING_TIME_FORMAT,
+            time.gmtime(end_processing_time - start_processing_time)
+        )
+        msg = ['Number of duplicates: {0}'.format(duplicates.value),
+               'Documents processed: ' + str(doc_len),
+               'Processing time: {0}'.format(total_time)]
```
Comment (justinask7) [inline corrector_module/opmon_corrector/corrector_batch.py:161]: Optional comment: f-strings could be used

File: corrector_module/opmon_corrector/database_manager.py [hunk 8]
```diff
@@ -248,22 +251,25 @@ def update_old_to_done(self, list_of_docs):
 
         return number_of_updated_docs
 
-    def check_if_hash_exists(self, doc_hash):
+    def check_clean_document_exists(self, x_request_id, document):
```
Comment (justinask7) [inline corrector_module/opmon_corrector/database_manager.py:254]: As discussed - we could start defining type hints

File: corrector_module/opmon_corrector/tests/fixtures/raw_messages_batch_2.json [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        "requestSize" : 917,
+        "requestAttachmentCount" : 0,
+        "succeeded" : false,
+        "faultCode" : "Server.ServerProxy.UnknownService",
+        "faultString" : "Unknown service: SERVICE:LTT/TEST/TEST2/SUB2/doesnotexist",
+        "xRequestId" : "a7351790-8fba-428d-852b-e4b6e8cbcca8",
+        "serviceType" : "WSDL",
+        "insertTime" : 1671090094.87872
+    }
+]
```
Comment (justinask7) [inline corrector_module/opmon_corrector/tests/fixtures/raw_messages_batch_2.json:60]: new line would be nice

**Review by justinask7 — APPROVED** [review]

File: corrector_module/opmon_corrector/corrector_batch.py (hunk 7 above)
Comment (wisecrow) [reply corrector_module/opmon_corrector/corrector_batch.py]: Good catch! F-strings added

File: corrector_module/opmon_corrector/database_manager.py (hunk 8 above)
Comment (wisecrow) [reply corrector_module/opmon_corrector/database_manager.py]: Basic type hints added

**Pull request merged**